### PR TITLE
Turned on noImplicitAny, strictNullChecks and strictPropertyInitialization for fast-colors

### LINKED ISSUE #5343: Refactor `fast-colors` to support tsconfig with `noImplicitAny`, `strictNullChecks` and `strictPropertyInitialization`
As part of #5303 we turned off `@typescript-eslint/typedef` globally as it was overwritten across multiple packages. As part of that, we turned on `noImplicitAny`, `strictNullChecks`, `strictPropertyInitialization` for all packages. Turning this on revealed several issues related to both tsconfig settings in `fast-colors` and so it was turned off explicitly for that package. The errors should be fixed and addressed as part of this issue and the code should be tested to ensure that there are no breaks or side effects of the change. This issue exists to track turning that back on and resolving the errors.

### PR DESCRIPTION
**Pull request by NiteshSethi09**
# Pull Request

## 📖 Description

Made changes in tsconfig.json and removed all the eslint errors. Made relevant changes and passed all the test cases. closes #5343 

## 👩‍💻 Reviewer Notes

I've used a non-null assertion operator for removing linting errors and made some necessary changes in the codebase.

### General

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: packages/utilities/fast-colors/src/color-converters.ts [hunk 1]
```diff
@@ -117,7 +117,7 @@ function calcRgbOverlay(
 export function calculateOverlayColor(
     rgbMatch: ColorRGBA64,
     rgbBackground: ColorRGBA64,
-    rgbOverlay: ColorRGBA64 = null
+    rgbOverlay: ColorRGBA64 = null!
```
Comment (nicholasrice) [inline packages/utilities/fast-colors/src/color-converters.ts:120]: I don't think this is the right fix. It should accept a null type and default to null

**Comment by nicholasrice** [comment]
Thanks @NiteshSethi09. I'm not sure this is the right approach for this task though, I think what we *want* when we implement this tsconfig change is to update the codebase so that those null values are accepted / known values where necessary, and the code avoids those cases. Adding null-assertions IMO defeats the point of explicit null checking.

File: packages/utilities/fast-colors/src/color-converters.ts (hunk 1 above)
Comment (NiteshSethi09) [reply packages/utilities/fast-colors/src/color-converters.ts:120]: `
export function calculateOverlayColor(
    rgbMatch: ColorRGBA64,
    rgbBackground: ColorRGBA64,
    rgbOverlay: ColorRGBA64 = null!
): ColorRGBA64 {
    let alpha: number = 0;
    let overlay: ColorRGBA64 = rgbOverlay;

    if (overlay !== null) {
        alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
    } else {
        overlay = new ColorRGBA64(0, 0, 0, 1);
        alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
        if (alpha <= 0) {
            overlay = new ColorRGBA64(1, 1, 1, 1);
            alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
        }
    }
    alpha = Math.round(alpha * 1000) / 1000;

    return new ColorRGBA64(overlay.r, overlay.g, overlay.b, alpha);
}
`

Comment (NiteshSethi09) [reply packages/utilities/fast-colors/src/color-converters.ts:120]: `
   export function calculateOverlayColor(
    rgbMatch: ColorRGBA64,
    rgbBackground: ColorRGBA64,
    rgbOverlay: ColorRGBA64 = null!
   ): ColorRGBA64 {
    let alpha: number = 0;
    let overlay: ColorRGBA64 = rgbOverlay;

    if (overlay !== null) {
        alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
    } else {
        overlay = new ColorRGBA64(0, 0, 0, 1);
        alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
        if (alpha <= 0) {
            overlay = new ColorRGBA64(1, 1, 1, 1);
            alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
        }
    }
    alpha = Math.round(alpha * 1000) / 1000;

    return new ColorRGBA64(overlay.r, overlay.g, overlay.b, alpha);
   }
`

I have found in the function definition is it is already checking for null values and performing operations based on the values. Let me know whats on your mind. I am open to your thoughts.

Comment (NiteshSethi09) [reply packages/utilities/fast-colors/src/color-converters.ts:120]: `
   export function calculateOverlayColor(
    rgbMatch: ColorRGBA64,
    rgbBackground: ColorRGBA64,
    rgbOverlay: ColorRGBA64 = null!
   ): ColorRGBA64 {
    let alpha: number = 0;
    let overlay: ColorRGBA64 = rgbOverlay;

    if (overlay !== null) {
        alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
    } else {
        overlay = new ColorRGBA64(0, 0, 0, 1);
        alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
        if (alpha <= 0) {
            overlay = new ColorRGBA64(1, 1, 1, 1);
            alpha = calcRgbOverlay(rgbMatch, rgbBackground, overlay);
        }
    }
    alpha = Math.round(alpha * 1000) / 1000;

    return new ColorRGBA64(overlay.r, overlay.g, overlay.b, alpha);
   }
`

I have found in the function definition is it is already checking for null values and performing operations based on the values. Let me know what's on your mind. I am open to your thoughts.

Comment (NiteshSethi09) [reply packages/utilities/fast-colors/src/color-converters.ts:120]: @nicholasrice shall we come to a discussion to close this issue?

**Comment by NiteshSethi09** [comment]
@nicholasrice I've made the changes as you mentioned in the upper comments. Please have a look and let me know if there's anything missing.

File: packages/utilities/fast-colors/src/color-blending.ts [hunk 2]
```diff
@@ -316,7 +320,7 @@ export function blend(
         case ColorBlendMode.Burn:
             return blendBurn(bottom, top);
         case ColorBlendMode.Color:
-            return blendColor(bottom, top);
+            return blendColor(bottom, top)!;
```
Comment (chrisdholt) [inline packages/utilities/fast-colors/src/color-blending.ts:323]: I don't think trailing exclamation point shouldn't be necessary if we're handling null as expected.

File: packages/utilities/fast-colors/src/color-converters.spec.ts [hunk 3]
```diff
@@ -106,10 +106,11 @@ describe("Color converter functions", () => {
                 data.rgba.a
             );
 
-            const hsl: ColorHSL = rgbToHSL(rgb);
-            expect(hsl.h).to.be.closeTo(data.hsl.h, testPrecision);
-            expect(hsl.s).to.be.closeTo(data.hsl.s, testPrecision);
-            expect(hsl.l).to.be.closeTo(data.hsl.l, testPrecision);
+            const hsl: ColorHSL | null = rgbToHSL(rgb);
+            expect(hsl).not.to.be.null;
+            expect(hsl!.h).to.be.closeTo(data.hsl.h, testPrecision);
```
Comment (chrisdholt) [inline packages/utilities/fast-colors/src/color-converters.spec.ts:111]: are we handling null in rgbToHSL?

Comment (NiteshSethi09) [reply packages/utilities/fast-colors/src/color-converters.spec.ts:111]: Yes, If you check the definition of rgbToHSL it checks in the very first line whether the rgb parameter is of null type or of `ColorRGBA64`. If a null type is found it returns null else proceed with the functionality. I'm also checking the return type of `hsl` variable is not null (on line number 110).

Would you like if I write a test case to check if it returns null on passing null?

File: packages/utilities/fast-colors/src/color-blending.ts (hunk 2 above)
Comment (NiteshSethi09) [reply packages/utilities/fast-colors/src/color-blending.ts:323]: Yes, I can remove that too. Removing the trailing exclamation point will lead the `blend` function to return `ColorRGBA64 | null`.

**Comment by janechu** [comment]
Closing this with the removal of the package see #6969.

**Pull request closed without merging**